### PR TITLE
[CMake] set OSX_ARCHITECTURES in `add_swift_target_library_single`

### PR DESCRIPTION
**Pull request by edymtt**
This way we configure correctly `libBlocksRuntime`, which is not using
`add_swift_target_library` (where the code for setting `OSX_ARCHITECTURES`
currently lives, see #38415 and #38956)

Addresses rdar://96469791

### CODE REVIEW

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 1]
```diff
@@ -1346,6 +1346,18 @@ function(add_swift_target_library_single target name)
...
     endif()
+
+    # In the past, we relied on unsetting globally
+    # CMAKE_OSX_ARCHITECTURES to ensure that CMake would
+    # not add the -arch flag
+    # This is no longer the case when running on Apple Silicon,
+    # when CMake will enforce a default (see
+    # https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5291)
+    set_property(TARGET ${target} PROPERTY OSX_ARCHITECTURES "${SWIFTLIB_SINGLE_ARCHITECTURE}")
+    if (target_static)
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:1357]: Should this perhaps be `if(TARGET ${target_static})`?  I'm worried that the string may be non-empty, which would result in this evaluating to true, even though the target has not been setup.

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 2]
```diff
@@ -1346,6 +1346,18 @@ function(add_swift_target_library_single target name)
       target_link_options(${target} PRIVATE
         "LINKER:-current_version,${SWIFT_COMPILER_VERSION}")
     endif()
+
+    # In the past, we relied on unsetting globally
+    # CMAKE_OSX_ARCHITECTURES to ensure that CMake would
+    # not add the -arch flag
+    # This is no longer the case when running on Apple Silicon,
+    # when CMake will enforce a default (see
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:1354]: Can you please reflow the text here?

**Review by compnerd — APPROVED** [review]

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 1 above)
Comment (edymtt) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake]: Good point -- implemented in 848a488b5d9

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 3]
```diff
@@ -1346,6 +1346,18 @@ function(add_swift_target_library_single target name)
       target_link_options(${target} PRIVATE
         "LINKER:-current_version,${SWIFT_COMPILER_VERSION}")
     endif()
+
+    # In the past, we relied on unsetting globally
+    # CMAKE_OSX_ARCHITECTURES to ensure that CMake
+    # would# not add the -arch flag. This is no longer
```
Comment (shahmishal) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:1352]:
```suggestion
    # would not add the -arch flag. This is no longer
```

Comment (edymtt) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake]: Thanks for noticing the typo -- this should be addressed in d03731bf6d4

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please test Apple Silicon

**Review by compnerd — APPROVED** [review]
Thanks!  I can't believe we didn't notice this earlier.

**Pull request merged**